features/mbedtls/targets/TARGET_STM/aes_alt.c

```diff
@@ -147,31 +150,93 @@ int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
-
-        /* Re-initialize AES IP with proper parameters */
-        if (HAL_CRYP_DeInit(&ctx->hcryp_aes) != HAL_OK)
-            return HAL_ERROR;
-        ctx->hcryp_aes.Init.OperatingMode = opmode;
-        ctx->hcryp_aes.Init.ChainingMode = CRYP_CHAINMODE_AES_CBC;
-        ctx->hcryp_aes.Init.KeyWriteFlag = CRYP_KEY_WRITE_ENABLE;
-        if (HAL_CRYP_Init(&ctx->hcryp_aes) != HAL_OK)
-            return HAL_ERROR;
-    }
+    /* At this moment only, we know we have CBC mode: Re-initialize AES
+        IP with proper parameters and apply key and IV for multi context usecase */
+    if (HAL_CRYP_DeInit(&ctx->hcryp_aes) != HAL_OK)
+        return HAL_ERROR;
+    ctx->hcryp_aes.Init.OperatingMode = opmode;
+    ctx->hcryp_aes.Init.ChainingMode = CRYP_CHAINMODE_AES_CBC;
+    ctx->hcryp_aes.Init.KeyWriteFlag = CRYP_KEY_WRITE_ENABLE;
+    if (HAL_CRYP_Init(&ctx->hcryp_aes) != HAL_OK)
+        return HAL_ERROR;
 
     status =  HAL_CRYPEx_AES(&ctx->hcryp_aes, input, length, output, 10);
 
     return status;
 }
+#else
+static int st_cbc_restore_context(mbedtls_aes_context *ctx){
+    uint32_t tickstart;
+    tickstart = HAL_GetTick();
+    while((ctx->hcryp_aes.Instance->SR & (CRYP_SR_IFEM | CRYP_SR_OFNE | CRYP_SR_BUSY)) != CRYP_SR_IFEM){
+        if ((HAL_GetTick() - tickstart) > ST_AES_TIMEOUT) {
+            return 1; // timeout: CRYP processor is busy
+        }
+    }
+    ctx->hcryp_aes.Instance->CR &= ~CRYP_CR_CRYPEN;
+    /* save initvector for multi-instance use of CRYP */
+    ctx->hcryp_aes.Instance->IV1RR = ctx->save_iv[3];
+    ctx->hcryp_aes.Instance->IV1LR = ctx->save_iv[2];
+    ctx->hcryp_aes.Instance->IV0RR = ctx->save_iv[1];
+    ctx->hcryp_aes.Instance->IV0LR = ctx->save_iv[0];
+    ctx->hcryp_aes.Phase = HAL_CRYP_PHASE_READY;
+    /* allow multi-instance of CRYP use: restore context for CRYP hw module */
+    ctx->hcryp_aes.Instance->CR = ctx->ctx_save_cr;
+    return 0;
+}
+static int st_cbc_save_context(mbedtls_aes_context *ctx){
+    uint32_t tickstart;
+    tickstart = HAL_GetTick();
+    while((ctx->hcryp_aes.Instance->SR & (CRYP_SR_IFEM | CRYP_SR_OFNE | CRYP_SR_BUSY)) != CRYP_SR_IFEM){
+        if ((HAL_GetTick() - tickstart) > ST_AES_TIMEOUT) {
+            return 1; // timeout: CRYP processor is busy
+        }
+    }
+    /* allow multi-instance of CRYP use: save context for CRYP HW module CR */
+    ctx->ctx_save_cr = ctx->hcryp_aes.Instance->CR;
+    ctx->hcryp_aes.Instance->CR &= ~CRYP_CR_CRYPEN;
+    /* save initvector for multi-instance use of CRYP */
+    ctx->save_iv[3] = ctx->hcryp_aes.Instance->IV1RR;
+    ctx->save_iv[2] = ctx->hcryp_aes.Instance->IV1LR;
+    ctx->save_iv[1] = ctx->hcryp_aes.Instance->IV0RR;
+    ctx->save_iv[0] = ctx->hcryp_aes.Instance->IV0LR;
+    if ((ctx->ctx_save_cr & CRYP_CR_CRYPEN) == CRYP_CR_CRYPEN) {
+        ctx->hcryp_aes.Instance->CR &= CRYP_CR_CRYPEN;
+    }
+    return 0;
+}
 #endif /* TARGET_STM32L486xG */
-
 int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
```

Comment:
It seems that procedure to save/restore contexts is quite involved and has different requirements for different functions. Would it be possible to include a comment somewhere in the code that documents in plain English what is the thinking behind these operations to enable interleaved uses of the accelerator with multiple contexts?